twa

```diff
@@ -222,7 +222,7 @@ function stage_2_security_headers {
   fi
 
   xfo=$(get_header "X-Frame-Options" <<< "${headers}")
-  xfo=${xfo,,}
+  xfo=`echo ${xfo} | tr '[:upper:]' '[:lower:]'`
```

Comment:
Per #11, `twa` is not going to support ancient `bash` versions -- there are too many subtle feature changes between 3 and 4, and the lack of nice builtins for string transformation is going to make future checks less flexible.

Reply:
You then lose all the OSX users. My customer base is 50% OSX by my logs.

Reply:
Based on the feedback in other issues/PRs, I don't think that this will be a huge issue: users who are technical enough to run `twa` can probably figure out how to install `bash` via Homebrew or another package manager.

Reply:
I disagree. Many people will run shell scripts, but not want to change bash, much less run brew or ports due to the other issues it causes. For the sake of 3 lines of code that are essentially similar, you are excluding a large user base. You could have a bash version detect routine at the beginning that switches in code, but seriously, for 3 lines, just make it backward compatible.

Reply:
> Many people will run shell scripts, but not want to change bash, much less run brew or ports due to the other issues it causes.

Based on the feedback in #1 and #7, I suspect that running a Homebrewed `bash` won't cause major issues for anybody -- [`bash` is one of the top 200 packages in Homebrew](https://formulae.brew.sh/analytics/install/365d/), and I (speaking as a Homebrew maintainer) haven't seen any substantive reports of breakage when people install it.

> For the sake of 3 lines of code that are essentially similar, you are excluding a large user base.

Per #1 (and the other closed issues that mention this), these three lines are a *symptom* of the problem, rather than the cause. `bash` 3.2 is *ancient*, and is only alive because Apple refuses to upgrade it (presumably because they're unwilling to distribute a GPLv3'd project). If `twa` assumes a base dependency of `bash` 3.2, the following features suddenly become unavailable for both new checks and refactoring current ones:

* Associative arrays
  * **Key**: At some point, `twa` should be refactored to use associative arrays -- HTTP headers map naturally to the notion of an associate array, and keeping all headers in a well-defined data structure will help avoid subtle string-parsing issues when extracting both headers and fields from those headers.
* All sorts of parameter expansion that could save `wc`, `tr`, `sed`, and `cut` processes: case processing, substring removal, find-and-replace, length counting, and so on.
  * **Key**: There are many parts of `twa` that currently use `wc`, `tr`, `sed`, and `cut` that I would like to refactor to use these features!

`bash` 4+ also subtly improves many common idioms: `set -e` correctly handles `[[ ]]` and `(( ))`, for example. I'd like to add `set -e` to `twa` in the near future, which will mean extra work handling the above case if `bash` 3.2 is settled upon.

In sum: it's really, *really* unfortunate that OS X continues to ship a version of `bash` that's more than a decade old. I use OS X professionally, and have dealt personally with scripts that don't behave well on OS X for various reasons (assuming a GNU userspace, assuming and not testing for `bash` 4, etc). It gets annoying fast. But that doesn't change the underlying calculus for maintainers: that `bash` 3.2 is functionally extinct, and that attempting to backport their scripts to 3.2 will only result in growing pains and compatibility hell when 3.2 and 4.x inevitably disagree on behavior. For those reasons, `twa` is never going to support 3.2.